scripts/build/gen_handles.py

```diff
@@ -36,7 +36,6 @@
 # This is needed to load edt.pickle files.
 sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..',
                                 'dts', 'python-devicetree', 'src'))
```

Comment:
All 4 of these lines (these 3 and the one below that is being removed) are part of the same change. If line 39 is not needed, then neither are lines 36-38.

I added these lines in because not having them was making twister crash in CI in mysterious ways. Are you sure about this? Can you provide more details about your investigation and its results?

Reply:
I created a script to do some downstream specific devicetree checks.  In my case, I was doing a dynamic import of edtlib like this:
```python
    zephyr_devicetree_path = (
        zephyr_base / "scripts" / "dts" / "python-devicetree" / "src"
    )

    # Add Zephyr's python-devicetree into the source path.
    site.addsitedir(zephyr_devicetree_path)

    edtlib_path = zephyr_devicetree_path / "devicetree" / "edtlib.py"
    spec = importlib.util.spec_from_file_location(edtlib_path.name, edtlib_path)
    edtlib = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(edtlib)

    with open(edt_pickle, "rb") as edt_file:
        edt = pickle.load(edt_file)

    try:
        named_gpios = edt.get_node("/named-gpios")
    except edtlib.EDTError:
        # If the named-gpios node doesn't exist, return success.
        return True
    except Exception:
       logging.error("Unexpected exception: %s",  sys.exc_info()[0])  
```

This code failed to catch `edtlib.EDTError`, and printed out `Unexpected exception: <class 'devicetree.edtlib.EDTError'>`.  By dropping the dynamic import and using `edtlib = inspect.getmodule(edt)` after `pickle.load()`, I was able to correctly catch `edtlib.EDTError`  exceptions.

If I skip the call to update `sys.path` with  `site.addsitedir`, I get the exception `ModuleNotFoundError: No module named 'devicetree'` from `pickle.load()`.

So the conclusion from @jackrosenthal and me was that the import of edtlib created a duplicate object, but the `sys.path` update was still required.

Reply:
That code looks similar to how west dynamically imports extension command modules, but I think it's not doing the right thing in this case.

I tried this:
https://gist.github.com/mbolivar-nordic/82a0586d68847cf70b0305fa2b5c3fac
and got:

```
new modules: ['_datetime', '_string', '_struct', 'atexit', 'base64', 'binascii', 'cython_runtime', 'datetime', 'devicetree', 'devicetree._private', 'devicetree.dtlib', 'devicetree.grutils', 'logging', 'math', 'string', 'struct', 'textwrap', 'threading', 'traceback', 'typing', 'typing.io', 'typing.re', 'yaml', 'yaml._yaml', 'yaml.composer', 'yaml.constructor', 'yaml.cyaml', 'yaml.dumper', 'yaml.emitter', 'yaml.error', 'yaml.events', 'yaml.loader', 'yaml.nodes', 'yaml.parser', 'yaml.reader', 'yaml.representer', 'yaml.resolver', 'yaml.scanner', 'yaml.serializer', 'yaml.tokens']
module=None
edtlib=<module 'edtlib.py' from '/home/mbolivar/zp/zephyr/scripts/dts/python-devicetree/src/devicetree/edtlib.py'>
edtlib_imported_directly=<module 'devicetree.edtlib' from '/home/mbolivar/zp/zephyr/scripts/dts/python-devicetree/src/devicetree/edtlib.py'>
```

Note that `sys.modules` does not contain an entry for `devicetree.edtlib` after line 18 where the importlib machinery has finished, but instead contains a module named 'edtlib.py'.

Contrast that with what happens when you import edtlib directly.

Something looks fishy in the importlib usage, no?

Reply:
(Note that the `edtlib_imported_directly` lines are likely relying on a `pip install -e zephyr_base/scripts/dts/python-devicetree` that I did previously)

Reply:
The `pickle.load()` seems to be picking up all needed modules while the dynamic import of edtlib doesn't add any modules.

```python
    zephyr_devicetree_path = (
        zephyr_base / "scripts" / "dts" / "python-devicetree" / "src"
    )
    # Add Zephyr's python-devicetree into the source path.
    site.addsitedir(zephyr_devicetree_path)
    
    modules_before = set(sys.modules)
    with open(edt_pickle, "rb") as edt_file:
        edt = pickle.load(edt_file)

    edtlib_from_pickle = inspect.getmodule(edt)
    modules_after_pickle = set(sys.modules)

    edtlib_path = zephyr_devicetree_path / "devicetree" / "edtlib.py"
    spec = importlib.util.spec_from_file_location(edtlib_path.name, edtlib_path)
    edtlib_from_import = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(edtlib_from_import)

    modules_after_import = set(sys.modules)

    print(f"edtlib_from_pickle={edtlib_from_pickle}")
    print(f'new pickle modules: {sorted(modules_after_pickle.difference(modules_before))}')
    print(f"edtlib_from_import={edtlib_from_import}")
    print(f'new import modules: {sorted(modules_after_import.difference(modules_after_pickle))}')
```

Results:
```
edtlib_from_pickle=<module 'devicetree.edtlib' from '/mnt/host/source/src/third_party/zephyr/main/scripts/dts/python-devicetree/src/devicetree/edtlib.py'>
new pickle modules: ['_datetime', '_yaml', 'base64', 'binascii', 'cython_runtime', 'datetime', 'devicetree', 'devicetree._private', 'devicetree.dtlib', 'devicetree.edtlib', 'devicetree.grutils', 'math', 'yaml', 'yaml.composer', 'yaml.constructor', 'yaml.cyaml', 'yaml.dumper', 'yaml.emitter', 'yaml.error', 'yaml.events', 'yaml.loader', 'yaml.nodes', 'yaml.parser', 'yaml.reader', 'yaml.representer', 'yaml.resolver', 'yaml.scanner', 'yaml.serializer', 'yaml.tokens']

edtlib_from_import=<module 'edtlib.py' from '/mnt/host/source/src/third_party/zephyr/main/scripts/dts/python-devicetree/src/devicetree/edtlib.py'>
new import modules: []
```

Reply:
That too seems to point to the importlib usage not producing the desired results, at least to me. Would you agree?

Reply:
I agree, something odd is happening with importlib.

The static import seems to work correctly and  doesn't appear to be required as the `pickle.load()` `inspect.get_module()` produces the correct object.

Snippet below. I hacked up gen_dts_cmake.py so I would readily have a valid edt in pickle format.
```python
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'python-devicetree',
                                'src'))
from devicetree import edtlib

def main():
    args = parse_args()

    with open(args.edt_pickle, 'rb') as f:
        edt = pickle.load(f)

    edtlib_pickle = inspect.getmodule(edt)

    if edtlib_pickle is edtlib:
        print("pickled edtlib matches imported edtlib")
    else:
        print("pickled edtlib and imported edtlib do not match")

    try:
        edt.get_node("/blah-blah")
    except edtlib_pickle.EDTError:
        print(f"Exception from pickle edtlib")
        sys.exit(1)
    except edtlib.EDTError:
        print(f"Exception from imported edtlib")
        sys.exit(1)
```

Output:
```
pickled edtlib matches imported edtlib
Exception from pickle edtlib
```

Lastly, if I swap the exception order - checking the imported `edtlib.EDTerror` first, then the output produces `Exception from imported edtlib`.

So my conclusion is that it's safe to drop the explicit `import edtlib` from scripts that use `pickle.load`.

Reply:
> So my conclusion is that it's safe to drop the explicit import edtlib from scripts that use pickle.load.

Sounds reasonable as long as the CI gremlins don't show up again!